openhtf/util/measurements.py

```diff
@@ -189,100 +204,115 @@ class MeasuredValue(object):
+    self.value_set = True
+    self.measurement.Validate(self.value)
 
-  def __iter__(self):  # pylint: disable=invalid-name
-    if self.num_dimensions:
-      return self.values.iteritems()
-    raise InvalidDimensionsError(
-        'Cannot iterate over undimensioned measurement.')
 
-  @classmethod
-  def ForMeasurement(cls, measurement):
-    """Create an unset MeasuredValue for this measurement."""
-    if measurement.dimensions:
-      return cls(measurement.name, len(measurement.dimensions))
-    else:
-      return cls(measurement.name, 0)
+class DimensionedMeasuredValue(object):
+  """Class encapsulating actual values measured.
 
-  def SetValue(self, value):
-    """Set the value for this measurement, with some sanity checks."""
-    if self.num_dimensions:
-      raise InvalidDimensionsError(
-          'Expected %s-dimensional coordinates, got dimensionless value %s' % (
-              self.num_dimensions, value))
+  See the MeasuredValue class docstring for more info.  This class provides a
+  dict-like interface for indexing into dimensioned measurements.
+  """
 
-    if self.value is not None:
-      _LOG.warning(
-          'Overriding previous measurement %s value of %s with %s',
-          self.name, self.value, value)
-    self.value = value
+  def __init__(self, name, num_dimensions, measurement):
+    assert num_dimensions > 0, 'Must have 1 or more dimensions'
+    self.name = name
+    self.num_dimensions = num_dimensions
+    self.measurement = measurement
+    self.value_dict = collections.OrderedDict()
+
+  def __iter__(self):  # pylint: disable=invalid-name
+    """Iterate over items, allows easy conversion to a dict."""
+    return self.value_dict.iteritems()
 
   def __setitem__(self, coordinates, value):  # pylint: disable=invalid-name
     coordinates_len = len(coordinates) if hasattr(coordinates, '__len__') else 1
     if coordinates_len != self.num_dimensions:
       raise InvalidDimensionsError(
           'Expected %s-dimensional coordinates, got %s' % (self.num_dimensions,
                                                            coordinates_len))
-    if coordinates in self.values:
+
+    # Wrap single dimensions in a tuple so we can assume value_dict keys are
+    # always tuples later.
+    if self.num_dimensions == 1:
+      coordinates = (coordinates,)
+
+    if coordinates in self.value_dict:
       _LOG.warning(
           'Overriding previous measurement %s[%s] value of %s with %s',
-          self.name, coordinates, self.values[coordinates], value)
-    self.values[coordinates] = value
+          self.name, coordinates, self.value_dict[coordinates], value)
+    self.value_dict[coordinates] = value
+    self.measurement.outcome = Outcome.PARTIALLY_SET
 
   def __getitem__(self, coordinates):  # pylint: disable=invalid-name
-    return self.values[coordinates]
+    # Let the AttributeError propagate as that's fairly intuitive to handle.
+    return self.value_dict[coordinates]
```

Comment:
you don't wrap single dimensions in tuples here, so you can set [1] and then fail on getting [1] later